apm-agent-core/src/main/java/co/elastic/apm/agent/util/BinaryHeaderMap.java

```diff
@@ -0,0 +1,223 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2020 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.util;
+
+import co.elastic.apm.agent.objectpool.Recyclable;
+import co.elastic.apm.agent.report.serialize.DslJsonSerializer;
+
+import javax.annotation.Nullable;
+import java.nio.Buffer;
+import java.nio.CharBuffer;
+import java.nio.charset.CoderResult;
+import java.util.ArrayList;
+import java.util.Iterator;
+
+/**
+ * A map (not implementing the java.util.Map interface) that only supports String-byte[] pair additions and iterations.
+ * The map doesn't allocate during addition or iteration.
+ * This map does not support any form of concurrency. It can be either be in a write mode (through its {@link #add}
+ * method) or read mode (through the {@link #iterator()} API) at a given time. Reads and writes must not be
+ * performed by more than one thread concurrently.
+ * <p>
+ * NOTE: this map does not guarantee visibility, therefore ensuring visibility when switching from read to write mode
+ * (or the other way around) is under the responsibility of the map's user.
+ */
+public class BinaryHeaderMap implements Recyclable, Iterable<BinaryHeaderMap.Entry> {
+    public static final int MAXIMUM_HEADER_BUFFER_SIZE = DslJsonSerializer.MAX_VALUE_LENGTH * 10;
+
+    private CharBuffer valueBuffer;
+    private final ArrayList<String> keys;
+    private int[] valueLengths;
+    private final NoGarbageIterator iterator;
+
+    public BinaryHeaderMap() {
+        valueBuffer = CharBuffer.allocate(64);
+        keys = new ArrayList<>(10);
+        valueLengths = new int[10];
+        iterator = new NoGarbageIterator();
+    }
+
+    public int size() {
+        return keys.size();
+    }
+
+    public boolean isEmpty() {
+        return size() == 0;
+    }
+
+    public boolean add(String key, byte[] value) throws InsufficientCapacityException {
+        int valuesPos = valueBuffer.position();
+        CoderResult coderResult = IOUtils.decodeUtf8Bytes(value, valueBuffer);
```

Comment:
I'm not sure it's a valid assumption that all or even most header values are UTF-8 encoded strings. Maybe fall back to hex-encoding if it's not UTF-8 data?

Reply:
It is certainly not the assumption. This is only to support recording of textual headers (as Kafka records only support binary). I prefer not to capture anything if it is not textual, and currently this is the indication I have to whether it is or not. 
The problem with hex-decoding, is that it always works, even if the value was not hex-encoded, so it would make no sense.